### ISSUE TITLE
[Bug][Regression] {{input}} helper does not support all HTML types

### ISSUE THREAD

**Issue by denniss17** [emberjs/ember.js#14651]
I ran into a strange bug when updating to ember 2.10.0:

## How to reproduce
Use `{{input type='date'}}` somewhere in your application.

## Expected behaviour
The rendered `input` element as type `date`.

## Real behaviour
The helper is rendered as an `input` element with type `text`. It did work in previous ember versions. 

## Test Case
I developed a simple test case in a bare ember-cli application, which can be found [here](https://github.com/denniss17/ember-input-helper-test/). The specific test case can be found [here](https://github.com/denniss17/ember-input-helper-test/blob/master/tests/integration/components/input-test.js), and is copied here:

```es6
import { moduleForComponent, test } from 'ember-qunit';
import hbs from 'htmlbars-inline-precompile';

moduleForComponent('input', 'Integration | Component | input', {
  integration: true
});

// Base case
test('it has the text type by default', function(assert) {
  this.render(hbs`{{input}}`);
  assert.equal(this.$('input').attr('type'), 'text');
});

// This one fails
test('it has the date type', function(assert) {
  this.render(hbs`{{input type='date'}}`);
  assert.equal(this.$('input').attr('type'), 'date');
});

// However, this one is succesfull
test('it has the number type', function(assert) {
  this.render(hbs`{{input type='number'}}`);
  assert.equal(this.$('input').attr('type'), 'number');
});

// I think this is as intended
test('it has the text type when type is random', function(assert) {
  this.render(hbs`{{input type='asdfasdfasdfasdfasdf'}}`);
  assert.equal(this.$('input').attr('type'), 'text');
});
```

Test results:
```
ok 1 PhantomJS 2.1 - JSHint | app.js: should pass jshint
ok 2 PhantomJS 2.1 - JSHint | helpers/destroy-app.js: should pass jshint
ok 3 PhantomJS 2.1 - JSHint | helpers/module-for-acceptance.js: should pass jshint
ok 4 PhantomJS 2.1 - JSHint | helpers/resolver.js: should pass jshint
ok 5 PhantomJS 2.1 - JSHint | helpers/start-app.js: should pass jshint
ok 6 PhantomJS 2.1 - Integration | Component | input: it has the text type by default
not ok 7 PhantomJS 2.1 - Integration | Component | input: it has the date type
    ---
        actual: >
            text
        expected: >
            date
        stack: >
            http://localhost:7357/assets/tests.js:133:17
            wrapper@http://localhost:7357/assets/test-support.js:7277:34
            runTest@http://localhost:7357/assets/test-support.js:3256:32
            run@http://localhost:7357/assets/test-support.js:3241:11
            http://localhost:7357/assets/test-support.js:3383:14
            process@http://localhost:7357/assets/test-support.js:3042:24
            begin@http://localhost:7357/assets/test-support.js:3024:9
            http://localhost:7357/assets/test-support.js:3084:9
        Log: |
    ...
ok 8 PhantomJS 2.1 - Integration | Component | input: it has the number type
ok 9 PhantomJS 2.1 - Integration | Component | input: it has the text type when type is random
ok 10 PhantomJS 2.1 - JSHint | integration/components/input-test.js: should pass jshint
ok 11 PhantomJS 2.1 - JSHint | resolver.js: should pass jshint
ok 12 PhantomJS 2.1 - JSHint | router.js: should pass jshint
ok 13 PhantomJS 2.1 - JSHint | test-helper.js: should pass jshint
```

## Remarks
It does also not seem to work for datetime and datetime-local. However, as can be seen in the testcase any random string is "normalized" to type `text`. I think somewhere the date types are missing as valid types.

Another strange observation is that it seems to work when running the tests in the chrome browser.

**Comment by Serabe** [comment]
The problem seems to be related to PhantomJS, since I cannot reproduce it in Chrome (and firefox does not implement date, so...)

**Comment by denniss17** [comment]
Yes, but it worked before in Ember, using the same PhantomJS version. So I'm not entirely sure if it is caused by Ember or PhantomJS

**Comment by denniss17** [comment]
I think I discovered the cause. The `canSetTypeOfInput` function in https://github.com/emberjs/ember.js/blob/v2.10.0/packages/ember-glimmer/lib/components/text_field.js#L14 checks if the type of the input field can be set to the given type. This is done by creating a "dummy" input element on which the type is set, after which the actual type is checked with the given type.

I copied the function over to the test case ([link](https://github.com/denniss17/ember-input-helper-test/commit/756fee6a91a777d6254c5bd381e51993d1632c7a)), added some HTML5 input types. This test case fails also on the 'date' type.

So the conclusion is that PhantomJS does not support this date type. However, because of the changed "type checking" in the new `ember-glimmer`, this is only discovered as of now. Maybe it is still an idea to change it, as I would expect the type of the input element to what I say it should be, independent of the support of the browser.

**Comment by rwjblue** [comment]
The same type checking was done in 2.9 and prior versions also ([see here](https://github.com/emberjs/ember.js/blob/v2.8.3/packages/ember-htmlbars/lib/components/text_field.js#L13)). I'm not sure why this would change in 2.10 (the type checking code seems to be identical.  What Ember version are you comparing against?

**Comment by Serabe** [comment]
The problem is jQuery takes the `type` value from the `attributes` collection, while `type` returns the type that the browser is actually rendering. Take a look at [this bin](http://jsbin.com/keruve/1/edit?js,console)

**Comment by denniss17** [comment]
Hmm, the function seems to be identical. The jQuery explantion makes sense, however I do not see JQuery being used apart from in the unit tests.

I created a branch of the same repo which uses ember 2.9.1, which can be found [here](https://github.com/denniss17/ember-input-helper-test/tree/ember-2.9.1). The [test results](https://travis-ci.org/denniss17/ember-input-helper-test/builds/180043658) show that the input type is set to whatever I set it to, but that the PhantomJS browser still doesn't support the date type

**Comment by pixelhandler** [comment]
@denniss17 The input helpers have had a long life I'm curious if that life is close to coming to an end. They are still based on original View objects not actual components. So essentially are helpers, not components. As a workaround I sometimes create a component that simple extends the Textfield view, e.g. https://github.com/pixelhandler/ember-jsonapi-resources-form/blob/master/addon/components/number-input.js#L27-L31

With the advances made in HTMLBars, Glimmer, and Glimmer 2 I'm curious if it's better to just use HTML input vs the TextField View object. I find the input helpers a bit archaic in some ways. How about making your own components and using tagName: "input" and set the type as you see fit?

**Comment by iezer** [comment]
I'm getting this same issue as well in phantomJS only, not in Chrome.

**Comment by denniss17** [comment]
This issue is around since ember 2.10 with the introduction of Glimmer. This is caused by PhantomJS not supporting native `date` input, falling back to a default `text` input. I think this is not particularly wrong, as you have to keep in mind that some browsers do not have datepickers.

In the end, we resolved to using a datepicker when date is not supported, and using a native datepicker if it is resolved. We adapted the tests to this (fill in date on native datepicker, otherwise use test helper of custom datepicker)

**Comment by nwhittaker** [comment]
Observing this problem in Safari 11.0.3 with Ember 3.0. This behavior gets in the way of attribute-based selections (i.e. `input[type=date]`) in JavaScript or CSS.

**Comment by Serabe** [comment]
[Safari has no support for input of type date](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date).

**Comment by nwhittaker** [comment]
>[Safari has no support for input of type date](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date).

I'd point out, per the Browser Compatibility foot notes for Safari, "It is recognized but there is no UI.[1]" I'm not entirely sure what exactly is supported, but presumably the base expectation here is that Ember remains unopinionated and respects whatever `type` attribute is provided (recognized or not). 

Given how html works contrasted with the current Ember documentation, I have no reason to think I couldn't invoke `{{input type='asdf'}}` in a template and select the resulting element with `input[type=asdf]`. Not that I'd do that, but it illustrates the point that html is an evolving vocabulary with well-defined fall-back behaviors that, in this specific case, Ember appears to be deviating from.

[1] https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date#Browser_compatibility

**Comment by pixelhandler** [comment]
@denniss17 @iezer @locks @nwhittaker @rwjblue is this still an issue, perhaps we should close or create a new reproduction of this, what do you think?
